Source/ServiceCall.cpp

```diff
@@ -106,13 +107,54 @@ static int handleGetParameters(void** params, int len){
       *value = settings.output_offset;
     }else if(strncmp(SYSEX_CONFIGURATION_OUTPUT_SCALAR, p, 2) == 0){
       *value = settings.output_scalar;
+    }else if(strncmp(SYSEX_CONFIGURATION_RESOURCE_COUNT, p, 2) == 0){
+      *value = (int)registry.getNumberOfResources();
+    }else if(strncmp(SYSEX_CONFIGURATION_RESOURCE_BY_NAME, p, 2) == 0 && len > index){
```

Comment:
let's not use GetParameters for retrieving a resource

Reply:
> let's remove GetParameter changes for now, focus on LoadResource service call

Are you saying that we should not have memory-mapped access to resources? This is a rather important feature for lookup tables.

> remove LoadResource operation name

How should the code in handleLoadResource be invoked in such case?

> using resources for Magus colours is a good way to show how resources can be used, but it's not a good feature to deploy since it only enables users to delete them. If in the future we create a tool for generating custom colours then we can bring this back.

Not only delete - it can be loaded from sysex and that's is the most convenient way to have custom colors installed. So just making a few sysex files available would be adequate. It will also be possible to make a patch that generates them when we'll have service calls for saving presets (which I think would be the next logical thing to work on). It was also possible in initial version before writing to flash was removed. Then only problem with this is that we can't make a tool that uses LED themselves as preview, only entering RGB values can be supported without access. But generating files would be a neat demo for preset saving.

Reply:
How can the patch find out how large a resource is?

Reply:
From resource header. So when you have memory-mapped access, you just check its size field. When it's not memory mapped, load call first fetches header separately, then it can fetch necessary amount of data (based on offset/length/total size). This part is done in OwlProgram repo.

Reply:
> Are you saying that we should not have memory-mapped access to resources? This is a rather important feature for lookup tables.

lookup tables will be retrieved in the same way they currently are, though the firmware side could be updated to look for a matching resource.

For other resource types I think LoadResource should work the way I've described elsewhere:
- in/out argument pointer to data pointer (ie void**)
- in/out argument pointer to length
On the first call the data pointer is NULL, length is set, and iff the resource is memory mapped then data pointer is also set.
Then there's no need for operation mode parameter either, right?

Reply:
We should only allow getting access to resource as both memorymapped and not memorymapped data in a single call if results are output as const data. If we want to allow mutable access, we would need a separate call that would be loading data even if it's available as memorymapped resource on flash. Otherwise a patch that works for one user could fail when loaded by another one due to different slot used for storing resource.

That's why I've made separate calls that can be used explicitly and we will always know if their results are mutable. And I was considering a loader class that would allow access to mutable or immutable data.

This issue is very real, because if we do it exactly the way you describe, we won't be able to tell if it's safe to modify data after the making resource load call. It may be read from flash. Not only that, but with QSPI flash everything would be available as memory mapped resource, so this fallback loading would be pointless. And with current firmware we don't have SPI NOR storage working, so nothing would be loaded to memory unless we allow user to force this.